### PR TITLE
Feature/reactor 347

### PR DESCRIPTION
**Pull request by silvexis**
Add CZRN formatted source events and the event_duration field

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -164,15 +148,3 @@ deploy: check-dependencies vendor                 ## wrap sls deploy
 
 remove:                                   ## call sls remove with correct args
 	@make action=remove sls
-
-
-# Uncomment if you know what you're doing. This is not really safe and may delete more tables than you intended.
```
Comment (killthrush) [inline Makefile]: we have a better way to handle this in core anyway

File: reactor/features/accounts/connect.py [hunk 2]
```diff
@@ -48,6 +48,14 @@ def connect_account(account_details, cfg, client_details, reactor_account_id):
     try:
         environment = env_dbh.get_from_account_id(account_id)
         environment_id = environment['environment_id']
+    except TypeError:
+        logger.error(f"This should never happen but I got none({environment}) trying to get_from_account_id "
+                     f"for account {account_id}")
+        logger.error(env_dbh.get_from_account_id.cache_info())
+        env_dbh.get_from_account_id.cache_clear()
+        # try again after clearing cache
+        environment = env_dbh.get_from_account_id(account_id)
```
Comment (killthrush) [inline reactor/features/accounts/connect.py]: hmm... sketchy, but at least we are logging it

File: reactor/features/ingest.py [hunk 3]
```diff
@@ -35,40 +35,74 @@ def ingest_events(events, namespace):
     """
     environments = dynamodb.CZReactorEnvironments()
     for event in events:
+        logger.debug(event)
```
Comment (killthrush) [inline reactor/features/ingest.py]: Should we remove line 38?

File: reactor/features/ingest.py [hunk 4]
```diff
@@ -35,40 +35,74 @@ def ingest_events(events, namespace):
...
-        's3_role': environment['account_details'].get('remote_cloudtrail_role'),
-        's3_external_id': environment['account_details'].get('remote_cloudtrail_id'),
-    }
-    spawn_lambdas_based_on_queue_size(reactor_namespace=namespace,
-                                      function_name='process-aws-cloudtrail-event',
-                                      parameters=lambda_function_parameters, messages_waiting=queue_msg_count)
+
+    sns_event_type, sns_event_data = identify_sns_event_type(event)
+
+    if sns_event_type == "EVENT_SNS_CLOUDTRAIL":
```
Comment (killthrush) [inline reactor/features/ingest.py]: I take it this solves the veracode problem?

File: reactor/features/ingest.py [hunk 5]
```diff
@@ -165,14 +199,13 @@ def process_aws_cloudtrail_event(queue_name, reactor_account_id, connected_accou
...
-                    'Unable to access bucket '
-                    '{} using role {} for account {}. Error: {}. This is likely because the permissions are bad. '
-                    'Will requeue event in the hope you can fix it'
-                ).format(bucket, role_arn, connected_account_id, str(err))
-
-                logger.error(message)
+                # If we made it here, most likely cause is bad permissions, but there is nothing we can do so
+                # we have to delete the message and move on.
+                message.delete()
+                logger.error(
```
Comment (killthrush) [inline reactor/features/ingest.py]: should this be a warning to match up with line 198?

File: reactor/features/resources/curate.py [hunk 6]
```diff
@@ -133,8 +132,10 @@ def process_new_resource(resource_id, timestamp, service_type, environment_id, d
 def write_resource_to_s3(resource_id, environment_id, resource):
     bucket_name = get_reactor_bucket_name()
     resource_id_hash = hashlib.sha256(resource_id.encode('UTF-8')).hexdigest()
-    # resource_object_key = f'resources/{environment_id}/{resource_id_hash}.gz'
-    resource_object_key = f'resources/{environment_id}/{resource_id_hash}.json'
+    resource_object_key = f'resources/{environment_id}/{resource_id_hash}.gz'
+    # resource_object_key = f'resources/{environment_id}/{resource_id_hash}.json'
```
Comment (killthrush) [inline reactor/features/resources/curate.py]: remove commented-out code here?

File: reactor/features/resources/curate.py [hunk 7]
```diff
@@ -147,24 +148,23 @@ def write_resource_to_s3(resource_id, environment_id, resource):
...
-                       Decimal(resource['first_seen']).quantize(Decimal(".00000"), rounding=ROUND_DOWN),
-                       resource['environment_id'],
-                       resource['cloud_account_id'],
-                       resource['service_type'],
-                       resource['disposition'],
-                       resource['count'],
-                       resource['raw_data'],
-                       resource['diff'],
-                       Decimal(resource['ttl']).quantize(Decimal(".00000"), rounding=ROUND_DOWN)]
+    # resource_values = [resource['res_id'],
```
Comment (killthrush) [inline reactor/features/resources/curate.py]: recommend nuking this commented-out code

File: reactor/features/resources/curate.py [hunk 8]
```diff
@@ -147,24 +148,23 @@ def write_resource_to_s3(resource_id, environment_id, resource):
...
+    #                    raw_data_json,
+    #                    diff_json,
+    #                    Decimal(resource['ttl']).quantize(Decimal(".00000"), rounding=ROUND_DOWN)]
     # NOTES: TSV expects a specific field order, only append new fields, otherwise downstream core mappings
     #        must be updated. In addition we round down to match the precision of the downstream table schema
-    # resource_object_value = utils.encode_and_compress("\t".join(map(str, resource_values)))
-    resource_object_value = "\t".join(map(str, resource_values)).encode('utf-8')
+    # resource_object_value = "\t".join(map(str, resource_values)).encode('utf-8')
 
     # JSON Format - Will be the preferred format once RedShift Spectrum supports JSON
```
Comment (killthrush) [inline reactor/features/resources/curate.py]: please update comment - we're effectively eliminating spectrum with this next release

File: reactor/utils/czrn.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2016-present, CloudZero, Inc. All rights reserved.
+# Licensed under the BSD-style license. See LICENSE file in the project root for full license information.
+import ipaddress
+import re
+
+
+def convert(input_string, service, region, account_id):
```
Comment (killthrush) [inline reactor/utils/czrn.py]: We really should have tests for this.  How sure are you that this actually works?

File: reactor/utils/czrn.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+    return f"czrn:{provider}:{service}:{region}:{account_id}:{network_id}:{subnet_id}:{interface_id}:" \
+           f"{resource_type}:{resource}"
+
+
+def is_valid_hostname(hostname):
+    if not hostname or len(hostname) > 255:
+        return False
+
+    if hostname[-1] == ".":
```
Comment (killthrush) [inline reactor/utils/czrn.py]: you could also do `hostname.rstrip('.')` unless you really would expect to see more than one dot

File: reactor/features/ingest.py (hunk 4 above)
Comment (silvexis) [reply reactor/features/ingest.py]: Yep

File: reactor/utils/czrn.py (hunk 9 above)
Comment (silvexis) [reply reactor/utils/czrn.py]: 100% It actually gets tested by the event normalization tests

Comment (killthrush) [reply reactor/utils/czrn.py]: Kewl

**Pull request merged**